compositor_pipeline/src/pipeline/output/whip.rs

```diff
@@ -0,0 +1,226 @@
+
+    #[error("Missing Host in endpoint URL")]
+    MissingHost,
+
+    #[error("Missing port in endpoint URL")]
+    MissingPort,
+
+    #[error("Failed to create RTC session description: {0}")]
+    RTCSessionDescriptionError(webrtc::Error),
+
+    #[error("Failed to set local description: {0}")]
+    LocalDescriptionError(webrtc::Error),
+
+    #[error("Failed to set remote description: {0}")]
+    RemoteDescriptionError(webrtc::Error),
+
+    #[error("Failed to parse {0} response body: {1}")]
+    BodyParsingError(String, reqwest::Error),
+
+    #[error("Failed to create offer: {0}")]
+    OfferCreationError(webrtc::Error),
+
+    #[error(transparent)]
+    PeerConnectionInitError(#[from] webrtc::Error),
+
+    #[error("Failed to convert ICE candidate to JSON: {0}")]
+    IceCandidateToJsonError(webrtc::Error),
+
+    #[error(transparent)]
+    SerdeJsonError(#[from] serde_json::Error),
+
+    #[error(transparent)]
+    PayloadingError(#[from] PayloadingError),
+}
+
+impl WhipSender {
+    pub fn new(
+        output_id: &OutputId,
+        options: WhipSenderOptions,
+        packets_receiver: Receiver<EncoderOutputEvent>,
+        request_keyframe_sender: Option<Sender<()>>,
+        pipeline_ctx: &PipelineCtx,
+    ) -> Result<Self, OutputInitError> {
+        let payloader = Payloader::new(options.video, options.audio);
+        let packet_stream = PacketStream::new(packets_receiver, payloader, 1400);
+
+        let should_close = Arc::new(AtomicBool::new(false));
+        let endpoint_url = options.endpoint_url.clone();
+        let bearer_token = options.bearer_token.clone();
+        let output_id = output_id.clone();
+        let should_close2 = should_close.clone();
+        let event_emitter = pipeline_ctx.event_emitter.clone();
+        let request_keyframe_sender = request_keyframe_sender.clone();
+        let tokio_rt = pipeline_ctx.tokio_rt.clone();
+
+        std::thread::Builder::new()
+            .name(format!("WHIP sender for output {}", output_id))
+            .spawn(move || {
+                let _span = span!(
+                    Level::INFO,
+                    "WHIP sender",
+                    output_id = output_id.to_string()
+                )
+                .entered();
+                start_whip_sender_thread(
+                    endpoint_url,
+                    bearer_token,
+                    should_close2,
+                    packet_stream,
+                    request_keyframe_sender,
+                    tokio_rt,
+                );
+                event_emitter.emit(Event::OutputDone(output_id));
+                debug!("Closing WHIP sender thread.")
+            })
+            .unwrap();
+
+        Ok(Self {
+            connection_options: options,
+            should_close,
+        })
+    }
+}
+
+impl Drop for WhipSender {
+    fn drop(&mut self) {
+        self.should_close
+            .store(true, std::sync::atomic::Ordering::Relaxed);
+    }
+}
+
+fn start_whip_sender_thread(
```

Comment:
```suggestion
fn run_whip_sender_thread(
```

start implies that you start the thread and exit immediately, run means that it will exit after all work is done